lib/problems.rb

```diff
@@ -1,13 +1,43 @@
 require_relative './stack.rb'
 
-# Time Complexity: ?
-# Space Complexity: ?
+# Time Complexity: O(n*m)
+# Space Complexity: O(m), but also potentially O(1)
 def balanced(string)
-  raise NotImplementedError, "Not implemented yet"
+  # assumes these are only valid chars
+  hash_table = {"}" => "{", "]" => "[", ")" => "("}  # space: O(m)? maybe O(1) bc size of this doesnt grow with input?
+  stack = Stack.new
+
+  string.each_char do |char|                    # time: O(n)
+    if hash_table.has_value?(char)              # time: O(m)?
```

Comment:
I don't think you need to use `has_value` here.  

I think better something like this

```ruby
matching_braces = {
  "}" => "{",
  ")" => "(",
  "]" => "[",
} 
stack = Stack.new
string.each_char do |char|
  if ! matching_braces[char]
    stack.push(char)
  elsif matching_braces[char] != stack.pop() 
    return false
  end
end
return stack.empty()
```